static/app/components/asyncComponentSearchInput.tsx

```diff
@@ -24,129 +46,156 @@ type DefaultProps = {
-     */
-    url: string;
-    /**
-     * Updates URL with search query in the URL param: `query`
-     */
-    updateRoute?: boolean;
-
-    onSearchSubmit?: (query: string, event: React.FormEvent) => void;
-    onSuccess: (data: object, resp: ResponseMeta | undefined) => void;
-    onError: () => void;
-
-    /**
-     * A render-prop child may be passed to handle custom rendering of the input.
-     */
-    children?: (otps: RenderProps) => React.ReactNode;
-  };
-
-type State = {
-  query: string;
-  busy: boolean;
-};
+  api: Client;
+  className?: string;
+  /**
+   * URL to make the search request to
+   */
+  url: string;
+  /**
+   * Updates URL with search query in the URL param: `query`
+   */
+  updateRoute?: boolean;
 
-/**
- * This is a search input that can be easily used in AsyncComponent/Views.
- *
- * It probably doesn't make too much sense outside of an AsyncComponent atm.
- */
-class AsyncComponentSearchInput extends React.Component<Props, State> {
-  static defaultProps: DefaultProps = {
-    placeholder: t('Search...'),
-    debounceWait: 200,
-  };
+  onSearchSubmit?: (query: string, event: React.FormEvent) => void;
+  onSuccess: (data: object, resp: ResponseMeta | undefined) => void;
+  onError: () => void;
+
+  /**
+   * A render-prop child may be passed to handle custom rendering of the input.
+   */
+  children?: (otps: RenderProps) => React.ReactElement;
+}
 
-  state: State = {
+function AsyncComponentSearchInput({
+  placeholder = t('Search...'),
+  debounceWait = 200,
+  api,
+  className,
+  url,
+  updateRoute,
+  onSearchSubmit,
+  onSuccess,
+  onError,
+  router,
+  location,
+  children,
+}: AsyncComponentSearchInputProps): React.ReactElement {
+  const [state, setState] = React.useState<{query: string; busy: boolean}>({
     query: '',
     busy: false,
-  };
-
-  immediateQuery = async (searchQuery: string) => {
-    const {location, api} = this.props;
-    this.setState({busy: true});
-
-    try {
-      const [data, , resp] = await api.requestPromise(`${this.props.url}`, {
-        includeAllArgs: true,
-        method: 'GET',
-        query: {...location.query, query: searchQuery},
-      });
-      // only update data if the request's query matches the current query
-      if (this.state.query === searchQuery) {
-        this.props.onSuccess(data, resp);
-      }
-    } catch {
-      this.props.onError();
-    }
-
-    this.setState({busy: false});
-  };
-
-  query = debounce(this.immediateQuery, this.props.debounceWait);
-
-  handleChange = (query: string) => {
-    this.query(query);
-    this.setState({query});
-  };
-
-  handleInputChange = (evt: React.ChangeEvent<HTMLInputElement>) =>
-    this.handleChange(evt.target.value);
+  });
+
+  // We need to use a mutable ref to keep reference to our latest query, else
+  // useCallback scope will always reference the previous value and our condition will always fail
+  const latestQuery = React.useRef<string>(state.query);
```

Comment:
```suggestion
  const latestQuery = React.useRef(state.query);
```

Reply:
Hmm, I like the explicit type. The nice thing about it is that as you read the useRef call, you immediately know that state.query is a string type and it doesnt require any action (e.g. hovering or clicking to go to type definition)

Reply:
I understand, but I think it adds extra complexity to the code that is not really needed. And in this case, because the `state.query` type was typed in the code above, we can see that this is of type `string` 🤷‍♀️

Reply:
@priscilawebdev I dont know if complexity is the right word here, the code does exactly the same with and without the type notation, but I would argue that readability improves with the explicit type. What you described by `because the state.query type was typed in the code above we can see that this has to be of type string` is exactly what I would want to avoid. If a new developer is reading this code, then to figure out that latestQuery is of type string, they either have to hover over latestQuery to get the type inference (assuming their editor supports that), or lookup state.query type wherever it was defined. In other words, just by reading `React.useRef<string>(state.query)` you can know exactly what the code does and no other action is necessary which is not the case when we do not add an explicit type.